### PR TITLE
Move MERGE checks for nulls in non nullable columns to planner

### PR DESCRIPTION
**Pull request by homar**
an improvement
> Is this a change to the core query engine, a connector, client library, or the SPI interfaces? (be specific)
core query engine
> How would you describe this change to a non-technical end user or system administrator?
It should not affect end user directly
## Related issues, pull requests, and links
Fixes: https://github.com/trinodb/trino/issues/13795

## Documentation

(x) No documentation is needed.
( ) Sufficient documentation is included in this PR.
( ) Documentation PR is available with #prnumber.
( ) Documentation issue #issuenumber is filed, and can be handled later.

## Release notes

(x) No release notes entries required.
( ) Release notes entries required with the following suggested text:

```markdown
# core
* Fix some things. ({issue}`issuenumber`)
```

### CODE REVIEW

**Comment by findepi** [comment]
> Homar/add check for nulls in nonnullable columns in planner

@homar can you please give the PR a name?

File: core/trino-main/src/main/java/io/trino/sql/planner/QueryPlanner.java [hunk 1]
```diff
@@ -698,7 +700,16 @@ public MergeWriterNode plan(Merge merge)
                     Expression original = mergeCase.getSetExpressions().get(index);
                     Expression setExpression = coerceIfNecessary(analysis, original, original);
                     subPlan = subqueryPlanner.handleSubqueries(subPlan, setExpression, analysis.getSubqueries(merge));
-                    rowBuilder.add(subPlan.rewrite(setExpression));
+                    Expression rewritten = subPlan.rewrite(setExpression);
+                    if (nonNullableColumnHandles.contains(dataColumnHandle)) {
+                        int fieldIndex = requireNonNull(mergeAnalysis.getColumnHandleFieldNumbers().get(dataColumnHandle), "Could not find fieldIndex for non nullable column");
+                        ColumnSchema columnSchema = dataColumnSchemas.get(fieldIndex);
+                        String columnName = columnSchema.getName();
+                        Expression coalesced = new CoalesceExpression(rewritten, new Cast(failFunction(metadata, session, INVALID_ARGUMENTS, "Assigning NULL to non-null MERGE target table column " + columnName), toSqlType(columnSchema.getType())));
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/sql/planner/QueryPlanner.java:708]: coalesced -> checked

or, I'd perhaps simply reassign `rewritten` and keep `rowBuilder.add(rewritten);` after the `if`.

File: core/trino-main/src/main/java/io/trino/sql/planner/QueryPlanner.java [hunk 2]
```diff
@@ -698,7 +700,16 @@ public MergeWriterNode plan(Merge merge)
...
+                        rowBuilder.add(coalesced);
+                    } else {
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/sql/planner/QueryPlanner.java]: newline before `else`

File: core/trino-main/src/main/java/io/trino/sql/planner/QueryPlanner.java [hunk 3]
```diff
@@ -707,7 +707,8 @@ public MergeWriterNode plan(Merge merge)
                         String columnName = columnSchema.getName();
                         Expression coalesced = new CoalesceExpression(rewritten, new Cast(failFunction(metadata, session, INVALID_ARGUMENTS, "Assigning NULL to non-null MERGE target table column " + columnName), toSqlType(columnSchema.getType())));
                         rowBuilder.add(coalesced);
-                    } else {
+                    }
+                    else {
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/sql/planner/QueryPlanner.java:711]: belongs to the first commit

File: core/trino-main/src/main/java/io/trino/sql/planner/QueryPlanner.java [hunk 4]
```diff
@@ -764,6 +775,7 @@ public MergeWriterNode plan(Merge merge)
                 subPlan.getRoot(),
                 projectionAssignmentsBuilder.build());
 
+
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/sql/planner/QueryPlanner.java]: redundant newline  (fixed in the second commit, please amend the first one)

File: core/trino-main/src/main/java/io/trino/sql/planner/QueryPlanner.java [hunk 5]
```diff
@@ -773,12 +785,12 @@ public MergeWriterNode plan(Merge merge)
                         .put(caseNumberSymbol, new SubscriptExpression(mergeRowSymbol.toSymbolReference(), new LongLiteral(Long.toString(rowType.getFields().size()))))
                         .build());
 
+
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/sql/planner/QueryPlanner.java]: redundant newline  (fixed in the second commit, please amend the first one)

**Review by findepi — APPROVED** [review]

File: core/trino-main/src/main/java/io/trino/sql/planner/QueryPlanner.java (hunk 3 above)
Comment (homar) [reply core/trino-main/src/main/java/io/trino/sql/planner/QueryPlanner.java:711]: strange that checkstyle didn't complain :/

Comment (homar) [reply core/trino-main/src/main/java/io/trino/sql/planner/QueryPlanner.java:711]: anyway there will be no else

Comment (findepi) [reply core/trino-main/src/main/java/io/trino/sql/planner/QueryPlanner.java]: > strange that checkstyle didn't complain :/

that's something we can try to improve. do you want to take a stab?

**Review by findepi — APPROVED** [review]

**Comment by homar** [comment]
Failure is not related, I opened an issue for it: https://github.com/trinodb/trino/issues/13849

**Comment by findepi** [comment]
> Failure is not related, I opened an issue for it: #13849

Or https://github.com/trinodb/trino/issues/13778.

**Comment by findepi** [comment]
Thanks!

**Pull request merged**